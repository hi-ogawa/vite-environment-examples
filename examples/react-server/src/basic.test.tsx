import { createManualPromise } from "@hiogawa/utils";
import { Window } from "happy-dom";
import React from "react";
import reactDomClient from "react-dom/client";
import { beforeAll, beforeEach, expect, test, vi } from "vitest";
import { initializeWebpackBrowser } from "./features/use-client/browser";

beforeAll(() => {
  // doing the same as examples/react-server/src/features/bootstrap/plugin.ts
  (globalThis as any).__raw_import = (id: string) =>
    import(/* @vite-ignore */ id);
});

// happy-dom
beforeEach(() => {
  const window = new Window({ url: "https://localhost:8080" });
  Object.assign(globalThis, {
    window,
    document: window.document,
  });
  return () => {
    window.close();
  };
});

async function testRender(container: reactDomClient.Container, page: string) {
  // react client browser
  initializeWebpackBrowser();
  const { default: reactServerDomClient } = await import(
    "react-server-dom-webpack/client.browser"
  );

  // fetch rsc stream via virtual module
  const testStream = await import("virtual:test-react-server-stream" + page);
  const testNode =
    reactServerDomClient.createFromReadableStream<React.ReactNode>(
      testStream.default,
    );

  // render
  const mounted = createManualPromise<void>();

  function Root() {
    React.useEffect(() => {
      mounted.resolve();
    }, []);
    return React.use(testNode);
  }

  reactDomClient.createRoot(container).render(<Root />);
  await mounted;
}

test("basic", async () => {
  await testRender(document.body, "/src/routes/page");
  await vi.waitUntil(() =>
    document.body.querySelector(`[data-hydrated="true"]`),
  );
  expect(document.firstElementChild).toMatchSnapshot();
});
