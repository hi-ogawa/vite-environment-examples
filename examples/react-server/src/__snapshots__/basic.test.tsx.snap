// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
<html>
  <head />
  <body>
    <div>
      <h4>
        Hello Server Component
      </h4>
      <div>
        Shared Component (
        server
        )
      </div>
      <div
        class="flex justify-center w-36 m-1 p-1 important:bg-[rgb(220,220,255)]"
      >
        unocss (server)
      </div>
      <div
        data-testid="server-action"
      >
        <h4>
          Hello Server Action
        </h4>
        <form
          action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        >
          <div>
            Count: 
            0
          </div>
          <button
            class="server-btn"
            name="value"
            value="-1"
          >
            -1
          </button>
          <button
            class="server-btn"
            name="value"
            value="1"
          >
            +1
          </button>
        </form>
        <form
          action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        >
          <h4>
            Hello useActionState
          </h4>
          <div
            style="display: flex; align-items: center; gap: 0.3rem;"
          >
            <div>
              1 + 1 = 
            </div>
            <input
              class="client-input"
              name="answer"
              placeholder="Answer?"
              required=""
            />
            <div
              data-testid="action-state"
            />
          </div>
        </form>
      </div>
      <div
        data-testid="client-component"
      >
        <h4>
          Hello Client Component
        </h4>
        <div>
          Shared Component (
          client
          )
        </div>
        <div
          class="flex justify-center w-36 m-1 p-1 important:bg-[rgb(255,220,220)]"
        >
          unocss (client)
        </div>
        <div>
          <div
            data-hydrated="true"
          >
            [hydrated: 
            true
            ]
          </div>
          <div>
            [effect: 
            <span>
              2
            </span>
            ]
          </div>
        </div>
        <div>
          Count: 
          0
        </div>
        <button
          class="client-btn"
        >
          -1
        </button>
        <button
          class="client-btn"
        >
          +1
        </button>
      </div>
    </div>
  </body>
</html>
`;

exports[`test async 1`] = `
<html>
  <head />
  <body>
    <div>
      <div>
        hello
      </div>
      <pre>
        {
  "name": "@hiogawa/vite-environment-examples-react-server",
  "private": true,
  "type": "module"
      </pre>
    </div>
  </body>
</html>
`;
